## Sklearn Gradient Boosting Regressor 
### We will peform GridSearch to find the optimal parameters to be used in the Gradient Boosting Regressor model to predict sales. Tuning of the hyperparameters are based on the MSE value obtained

In [1]:
#Importing relevant libraries
from data import train
from sklearn.svm import SVR
import numpy as np
from multiprocessing import cpu_count
from utils import StandardizedGridSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error as MSE
from sklearn.model_selection import train_test_split

In [2]:
#Separating predictors from sales column
X=train.drop('sales', axis=1)
y=train.sales

### We will first fit the training data into the default Gradient Boosting Regressor model to see the MSE value before conducting GridSearch to find the optimal hyperparameters

In [7]:
# conducting a separate train test split to fit into default Gradient Boosting Regressor Model

X_train,X_test,y_train,y_test = train_test_split(X,y, test_size=0.2,random_state =20)

In [8]:
#default for Gradient Boosting Regressor 
from sklearn.ensemble import GradientBoostingRegressor
GB_default = GradientBoostingRegressor()
GB_default.fit(X_train,y_train)


GradientBoostingRegressor()

In [9]:
# MSE value for Gradient Boosting Regressor default model
GBR_default_MSE=np.mean((y_test - GB_default.predict(X_test))**2)

GBR_default_MSE

2886.4762722716077

### We will now conduct GridSearchCV to see if a better model can be obtained

In [10]:
#Retriving saved Decision Tree Models
# Gradient Boosting Regressor Run 1

GB_run1=StandardizedGridSearchCV.load('models/Boosting9.p')
GB_run1

GridSearchCV(cv=10, estimator=GradientBoostingRegressor(), n_jobs=1,
             param_grid={'ccp_alpha': (0.01, 0.1, 1),
                         'learning_rate': (0.05, 0.1, 1), 'max_depth': [10],
                         'max_features': (None, 'auto'), 'n_estimators': [50]},
             refit='neg_mean_squared_error', return_train_score=True,
             scoring=['neg_mean_squared_error', 'r2'], verbose=10)

In [11]:
# Top 10 model results for Gradient Boosting Regressor Run 1

GB_run1.results.head(10)

50                0.035107   
          0.05          10        NaN          50                0.044803   
0.10      0.10          10        auto         50                0.110265   
                                  NaN          50                0.137082   
1.00      0.05          10        auto         50                0.046751   
0.01      0.10          10        auto         50                0.037861   
          0.05          10        auto         50                0.019256   
                                  NaN          50                0.041416   

                                                             mean_score_time  \
ccp_alpha learning_rate max_depth max_features n_estimators                    
1.00      0.10          10        NaN          50                   0.002802   
0.01      0.10          10        NaN          50                   0.003296   
1.00      0.10          10        auto         50                   0.003398   
          0.05          10        NaN          50                   0.003095   
0.10      0.10          10        auto         50                   0.003584   
                                  NaN          50                   0.003697   
1.00      0.05          10        auto         50                   0.003192   
0.01      0.10          10        auto         50                   0.004184   
          0.05          10        auto         50                   0.003289   
                                  NaN          50                   0.003290   

                                                             std_score_time  \
ccp_alpha learning_rate max_depth max_features n_estimators                   
1.00      0.10          10        NaN          50                  0.000404   
0.01      0.10          10        NaN          50                  0.000645   
1.00      0.10          10        auto         50                  0.001025   
          0.05          10        NaN          50                  0.000298   
0.10      0.10          10        auto         50                  0.000660   
                                  NaN          50                  0.000638   
1.00      0.05          10        auto         50                  0.000399   
0.01      0.10          10        auto         50                  0.001324   
          0.05          10        auto         50                  0.000459   
                                  NaN          50                  0.000456   

                                                             split0_test_neg_mean_squared_error  \
ccp_alpha learning_rate max_depth max_features n_estimators                                       
1.00      0.10          10        NaN          50                                  -1007.148977   
0.01      0.10          10        NaN          50                                  -1075.712130   
1.00      0.10          10        auto         50                                  -1016.489230   
          0.05          10        NaN          50                                  -1147.126324   
0.10      0.10          10        auto         50                                  -1034.271474   
                                  NaN          50                                  -1118.607157   
1.00      0.05          10        auto         50                                  -1164.832490   
0.01      0.10          10        auto         50                                  -1098.211643   
          0.05          10        auto         50                                  -1300.842633   
                                  NaN          50                                  -1317.580421   

                                                             split1_test_neg_mean_squared_error  \
ccp_alpha learning_rate max_depth max_features n_estimators                                       
1.00      0.10          10        NaN          50                                  -1488.989788   
0.01      0.10          10        

In [12]:
#Printing top 10 models for Gradient Boosting Regressor Run 1 based on test MSE value

GB_run1.results['mean_test_neg_mean_squared_error'].head(10)

ccp_alpha  learning_rate  max_depth  max_features  n_estimators
1.00       0.10           10         NaN           50             -2414.730690
0.01       0.10           10         NaN           50             -2476.764123
1.00       0.10           10         auto          50             -2500.674680
           0.05           10         NaN           50             -2519.584491
0.10       0.10           10         auto          50             -2521.825112
                                     NaN           50             -2535.384604
1.00       0.05           10         auto          50             -2552.715643
0.01       0.10           10         auto          50             -2556.335300
           0.05           10         auto          50             -2575.678639
                                     NaN           50             -2577.167875
Name: mean_test_neg_mean_squared_error, dtype: float64

In [13]:
#We aggreagted the train and test MSE scores of run1 to see if learning rate had a significant effect on the MSE values

(-GB_run1.results[['mean_train_neg_mean_squared_error', 'mean_test_neg_mean_squared_error']]).groupby('learning_rate').mean()

mean_train_neg_mean_squared_error  \
learning_rate                                      
0.05                                  108.105028   
0.10                                   36.145153   
1.00                                   44.441383   

               mean_test_neg_mean_squared_error  
learning_rate                                    
0.05                                2573.163968  
0.10                                2500.952418  
1.00                                3022.182704

### From the table above, we can see that learning_rate has a significant impact on the MSE Value and we will conduct GridSearchCV on a wider range of learning_rate values to find the best value

In [15]:
#We aggreagted the train and test MSE scores of run1 to see if ccp_alpha had a significant effect on the MSE values

(-GB_run1.results[['mean_train_neg_mean_squared_error', 'mean_test_neg_mean_squared_error']]).groupby('ccp_alpha').mean()

mean_train_neg_mean_squared_error  mean_test_neg_mean_squared_error
ccp_alpha                                                                     
0.01                               27.512708                       2727.591215
0.10                               40.454289                       2723.417936
1.00                              120.724568                       2645.289939

In [16]:
#Retriving saved Decision Tree Models
# Gradient Boosting Regressor Run 2

GB_run2=StandardizedGridSearchCV.load('models/Boosting10.p')
GB_run2

GridSearchCV(cv=10, estimator=GradientBoostingRegressor(), n_jobs=1,
             param_grid={'ccp_alpha': (0.01, 0.1, 1),
                         'learning_rate': array([0.005     , 0.00693878, 0.00887755, 0.01081633, 0.0127551 ,
       0.01469388, 0.01663265, 0.01857143, 0.0205102 , 0.02244898,
       0.02438776, 0.02632653, 0.02826531, 0.03020408, 0.03214286,
       0.03408163, 0.03602041, 0.03795918, 0.03989796, 0.04183673,
       0.04377551,...
       0.06316327, 0.06510204, 0.06704082, 0.06897959, 0.07091837,
       0.07285714, 0.07479592, 0.07673469, 0.07867347, 0.08061224,
       0.08255102, 0.0844898 , 0.08642857, 0.08836735, 0.09030612,
       0.0922449 , 0.09418367, 0.09612245, 0.09806122, 0.1       ]),
                         'max_depth': [10], 'max_features': (None, 'auto'),
                         'n_estimators': [50]},
             refit='neg_mean_squared_error', return_train_score=True,
             scoring=['neg_mean_squared_error', 'r2'], verbose=10)

In [17]:
# Top 10 results for Gradient Boosting Regressor Run 2
GB_run2.results.head(10)

50                0.051624   
          0.082551      10        NaN          50                0.108088   
          0.076735      10        auto         50                0.049977   
          0.070918      10        auto         50                0.030873   
          0.100000      10        auto         50                0.085106   
          0.074796      10        NaN          50                0.030507   
          0.076735      10        NaN          50                0.032781   
          0.067041      10        NaN          50                0.021151   

                                                             mean_score_time  \
ccp_alpha learning_rate max_depth max_features n_estimators                    
1.0       0.090306      10        NaN          50                   0.003598   
          0.074796      10        auto         50                   0.003997   
          0.092245      10        NaN          50                   0.003981   
          0.082551      10        NaN          50                   0.004089   
          0.076735      10        auto         50                   0.003383   
          0.070918      10        auto         50                   0.003791   
          0.100000      10        auto         50                   0.003485   
          0.074796      10        NaN          50                   0.003590   
          0.076735      10        NaN          50                   0.003387   
          0.067041      10        NaN          50                   0.003887   

                                                             std_score_time  \
ccp_alpha learning_rate max_depth max_features n_estimators                   
1.0       0.090306      10        NaN          50                  0.000497   
          0.074796      10        auto         50                  0.001610   
          0.092245      10        NaN          50                  0.000770   
          0.082551      10        NaN          50                  0.001041   
          0.076735      10        auto         50                  0.000487   
          0.070918      10        auto         50                  0.000884   
          0.100000      10        auto         50                  0.000492   
          0.074796      10        NaN          50                  0.000489   
          0.076735      10        NaN          50                  0.000500   
          0.067041      10        NaN          50                  0.001369   

                                                             split0_test_neg_mean_squared_error  \
ccp_alpha learning_rate max_depth max_features n_estimators                                       
1.0       0.090306      10        NaN          50                                  -1024.567493   
          0.074796      10        auto         50                                  -1039.739352   
          0.092245      10        NaN          50                                   -947.318615   
          0.082551      10        NaN          50                                   -985.932004   
          0.076735      10        auto         50                                   -992.457689   
          0.070918      10        auto         50                                  -1019.706984   
          0.100000      10        auto         50                                   -987.019454   
          0.074796      10        NaN          50                                  -1040.652505   
          0.076735      10        NaN          50                                  -1032.308355   
          0.067041      10        NaN          50                                  -1105.897619   

                                                             split1_test_neg_mean_squared_error  \
ccp_alpha learning_rate max_depth max_features n_estimators                                       
1.0       0.090306      10        NaN          50                                  -1534.213518   
          0.074796      10        

In [18]:
#Printing top 10 models for Gradient Boosting Regressor Run 2 based on test MSE value

GB_run2.results['mean_test_neg_mean_squared_error'].head(10)

ccp_alpha  learning_rate  max_depth  max_features  n_estimators
1.0        0.090306       10         NaN           50             -2380.512852
           0.074796       10         auto          50             -2383.715041
           0.092245       10         NaN           50             -2396.952036
           0.082551       10         NaN           50             -2404.009310
           0.076735       10         auto          50             -2406.116954
           0.070918       10         auto          50             -2412.790000
           0.100000       10         auto          50             -2414.845873
           0.074796       10         NaN           50             -2416.821154
           0.076735       10         NaN           50             -2418.580002
           0.067041       10         NaN           50             -2421.494067
Name: mean_test_neg_mean_squared_error, dtype: float64

### Comparing the 2 GridSearchCV models as well as the default model, we see that GB_run2 is the best model as it gives the lower test MSE score of 2380.51

In [19]:
## Printing best parameters from GB_run2

GB_run2.best_estimator_

GradientBoostingRegressor(ccp_alpha=1, learning_rate=0.09030612244897959,
                          max_depth=10, n_estimators=50)